mycpp/leaky_mylib.cc

```diff
@@ -137,51 +147,73 @@ bool CFileWriter::isatty() {
 //
 
 void Buf::Extend(Str* s) {
-  int n = len(s);
+  const int n = len(s);
 
-  assert(cap_ >= len_);
-  if (cap_ < len_ + n) {
-    cap_ = std::max(cap_ * 2, len_ + n);
-  }
-  // +1 for NUL.  TODO: consider making it a power of 2
-  data_ = static_cast<char*>(realloc(data_, cap_ + 1));
+  assert(cap_ >= len_ + n);
 
   memcpy(data_ + len_, s->data_, n);
   len_ += n;
   data_[len_] = '\0';
 }
 
 void Buf::Invalidate() {
-  free(data_);
+  // free(data_);
   len_ = -1;
   cap_ = -1;
-  data_ = nullptr;
+  // data_ = nullptr;
 }
 
 //
 // BufWriter
 //
 
+void BufWriter::ExpandBufCapacity(int n) {
+  if (!buf_) {
+    buf_ = NewBuf(n);
+    return;
+  }
+
+  assert(buf_->cap_ >= buf_->len_);
+
+  if (buf_->cap_ < buf_->len_ + n) {
+    // buf_->cap_ = std::max(buf_->cap_ * 2, buf_->len_ + n);
+
+    // +1 for NUL.  TODO: consider making it a power of 2
+    // buf_->data_ = static_cast<char*>(realloc(buf_->data_, buf_->cap_ + 1));
+
+    auto* b = NewBuf(std::max(buf_->cap_ * 2, buf_->len_ + n));
+    memcpy(b->data_, buf_->data_, buf_->len_);
+    b->len_ = buf_->len_;
+    b->data_[b->len_] = '\0';
+    // free(buf_->data_);
+    buf_ = b;
+  }
+}
+
+void BufWriter::Extend(Str* s) {
+  ExpandBufCapacity(len(s));
+  buf_->Extend(s);
```

Comment:
It seems like both of these lines do a memcpy() ?  Is that necessary?

What about `MarkSweepHeap::Reallocate()` which calls `realloc()`?